### ISSUE TITLE
Add incremental builds to build process

### ISSUE THREAD

**Issue by webinista** [department-of-veterans-affairs/vets-website#2753]
Primarily affects rx, hca projects.

**Comment by alexose** [comment]
@robbiethegeek @crwallace 

Quick update on this front (and please feel free to chime in!)

Incremental builds already happen in `npm run watch` mode, thank's to webpack's in-memory cache.  However, shutting down webpack always results in losing the cache.  Any time we rebuild (on the CI servers, for instance), we rebuild from scratch.

There are people working on persisting webpack's cache to disk, most notably https://github.com/mzgoddard/hard-source-webpack-plugin.  I'm not sure it's production-ready, but I am messing with it over here: #3074 

Long story short, we could save this along with the rest of the build cache and shave about 200 seconds off of each build (I think). 

E2E tests are the next big chunk.  I think we solve that problem by only running them on merge to staging, but that's a separate PR.

**Comment by jcmeloni** [comment]
@robbiethegeek @crwallace Do we still need to do work on incremental builds?

**Comment by alexose** [comment]
Yes, I would suspect we'll be getting to this in the next week or two.

**Comment by aub** [comment]
This would help out locally but wouldn't affect build times on Travis/Jenkins right, since they start from scratch each time?

**Comment by alexose** [comment]
The big win would be adding this plugin and caching its results with Travis's build cache feature.

Where it gets tricky is figuring out exactly when to invalidate it.  There are a few moving parts, and it requires some understanding of what webpack is doing behind the scenes...

I was thinking about a naive approach like "see if it builds successfully from cache, if not, try again from scratch".

**Comment by aub** [comment]
That makes me pretty nervous for a production build. Presumably it could fail to invalidate something that changes the site in some bad way but still allows the build to pass right?

**Comment by alexose** [comment]
Well, it would be relying on the tests to catch breaking changes.  I would be fine with turning off all caching for production builds, though, since it's really the builds within PRs that are the big pain point.

**Comment by robbiethegeek** [comment]
I think that doing this locally would have an impact on developer happiness, but if we can't affect dev and prod builds I think the ROI isn't high enough to prioritize before other build performance issues.

**Comment by jcmeloni** [comment]
@robbiethegeek I'm going to close this rather than move it over to vets-gov.team, as the build changes overall are going to be good enough and ROI decreases significantly after that. Reopen and tag me to move if you disagree!
